### ISSUE TITLE
Use common command detection routines

### ISSUE THREAD

**Issue by philwhineray** [firehol/firehol#99]
I have made what I think are the final updates to the firehol script to only use bash builtins or detected commands. I updated the unit test to help enforce it.

The user is able to specify commands to use by setting variables in `/etc/firehol/firehol-defaults.conf`. @ktsaou , when I update the other scripts, should they also read this file? I think this would be better than having many files.

It does mean they will need to understand or ignore the mark definitions. In the case of fireqos it could either keep using the generated values or it could do the same calculations as firehol (so the order of running would no longer matter but the fireqos values would reflect the config not necessarily the running firewall).

**Comment by ktsaou** [comment]
They can read this file and ignore the variables they don't need.

About the marks, when I designed the whole thing, I was expecting all the other tools (except firehol itself) to ignore the mark definitions in this file. FireHOL dynamically assigns marks (check synproxy, tproxy, etc). So all other tools should use the final mark definitions saved by firehol and ignore the ones defined in firehol-defaults.

I know there are a few folks that are using only FireQOS for example. These guys should use `rawmark` to match their own marks. Probably, if the firehol saved file does not exist, all tools except firehol should issue a warning if the `mark` match is being used.

So, in general, I suggest all tools except firehol, to have a stub `markdef`.

**Comment by philwhineray** [comment]
OK, I have converted fireqos now.

Hopefully everything is OK this time but some more testing would be a good idea, since it is quite a big change.

I have updated the `pre-commit` script to include tests for fireqos and now the configure.ac too, so hopefully everything is covered.

**Comment by ktsaou** [comment]
excellent! I will test it tomorrow.

**Comment by ktsaou** [comment]
Nice work!

I found the following minor issues:

1. `get_version()` is using `$SED_CMD` before commands management.
2. there is a `TPUT_COMMAND` instead of a `$TPUT_CMD` at line 297.

**Comment by philwhineray** [comment]
Thanks. I pushed fixes for these and also a typo that has been in the function for some time, and that firehol was still using `sed` direct.

I have also been updating the unit test infrastructure to detect and use [kcov](https://github.com/SimonKagstrom/kcov), so it should be easier to ensure as much as possible is checked. The current version still has some problems managing all the syntax of firehol but I am going to submit a patch which resolves most of that.

**Comment by ktsaou** [comment]
Nice!

Phil you have to give me a webminar for this excellent job you do with the infrastructure... I strangle to follow you (am I too old? :-)

**Comment by philwhineray** [comment]
I doubt you are too old :) more likely my powers of documentation are not as good as they should be.

We can try a webinar sometime but maybe not for a few weeks if you can wait? Meantime I am happy to try to explain whatever you would like to know more about.

Anyhow, I have just pushed some more commits.

The key change is that link-balancer now uses the same functions as firehol and fireqos (well, in addition it exports the variables it autodetects so that it does not re-detect them when it runs itself). So this could do with some more testing, I have done only a basic run which I added to the unit tests.

I also made a couple of fixes for other programs I noticed.

Finally, I have made the unit tests run without needing root at all; instead they rely on having user namespaces enabled for regular users. On some systems this is disabled by default so you will get prompted. It was much harder than it should have been due to this:

* http://lists.linuxfoundation.org/pipermail/containers/2014-June/034682.html
* https://lists.linuxcontainers.org/pipermail/lxc-users/2014-November/008099.html

Adding the fact one cannot bind mount over something under `/proc/net/` from shell commands - it always affects the running pid (of the mount command...) and I had to rebuild a fake `/proc` to get it to work. Maybe I will try my hand at kernel development next...

**Comment by ktsaou** [comment]
Nice!

Phil, please add to the pre-commit checks also these cases: X_CMD without $ or without $ and {}

In FireQOS there is a TPUT_CMD like this:

`$(TPUT_CMD colors 2>/dev/null)`

which should be:

`$($TPUT_CMD colors 2>/dev/null)`

So, it would be good if it could detect these cases too.

**Comment by philwhineray** [comment]
OK, it makes the check now and I fixed fireqos.in and link-balancer.in

**Comment by philwhineray** [comment]
Also the pre-commit script checks itself now and will not let you commit until the one in `.git/hooks` matches the one in `packaging`, to ensure the latest one is being used.

**Comment by ktsaou** [comment]
Thanks!

For some reason pre-commit does not allow me to commit, without error.
I only get this:

```txt
dpkg-gencontrol: warning: File::FcntlLock not available; using flock which is not NFS-safe
```

which I guess should be ignored.

**Comment by philwhineray** [comment]
Strange. It does not sound fatal though.

I would ignore it, or maybe install the perl module in question to silence it.

**Comment by ktsaou** [comment]
Phil, the issue is that the pre-commit script it does not allow me to commit. `git` exits without accepting the commit and without error.

I have to remove the pre-commit script, to commit anything.

I'll give it a look and let you know...

**Comment by philwhineray** [comment]
OK, I fixed it. I thought just the status of dpkg-gencontrol was checked, but it seems the output is too.

**Comment by philwhineray** [comment]
Also, I just added a commit 1ea9a58bd42be0770f611a60ba057484e4ac753e to a branch, to convert update-ipsets.in (including the changes you just made).

This one was much harder than the others, which is why I did not put it on master yet. Can you take a look when you have time, before I merge it.

**Comment by ktsaou** [comment]
yes of course.
I will do another commit in 2 mins, with some minor fixes.

**Comment by ktsaou** [comment]
done it

**Comment by philwhineray** [comment]
Thanks. I have also merged your update into the same update-ipsets-commands branch, so it uses $SED_CMD and $TR_CMD. It is now commit 1e5fa7befaec75c4330020ecbdda8123e8bc9994 overall.

**Comment by ktsaou** [comment]
@philwhineray I made a few updates to update-ipsets.
I hope I haven't destroyed the work you haven't committed yet.

**Comment by philwhineray** [comment]
I have been tracking the updates - it will mean a bit of rework wherever you changed something I also updated but it should not be huge, git seems pretty good at managing that sort of thing.

**Comment by philwhineray** [comment]
OK, I fixed it up again, hopefully I caught everything. Take a look at the branch update-ipsets-commands for the current state. If you prefer, I can do a rebase so it looks like a single commit on top of your work.

**Comment by ktsaou** [comment]
Yes please do so... I am lost trying to figure how to get a diff with master...

**Comment by philwhineray** [comment]
> Yes please do so... I am lost trying to figure how to get a diff with master...

I won't get a chance until tonight; maybe try this in the meantime to
get a diff on your local machine:

From your firehol repo, ensure you have everything either committed
or stashed, then:

~~~~
git checkout update-ipsets-commands
~~~~

This should make a local copy of the branch and associate it to the
upstream branch, in case you want to push anything back.

Then you can switch to and from, e.g. to go back to master:

~~~~
git checkout master
~~~~

To get updates from upstream, you need to `git pull` when you are
on the appropriate branch. See where you are with:

~~~~
git branch
~~~~

For diffs, you can specify a branch anywhere you can specify a commit,
so to diff the changes I made to the script, after getting a local
copy of the branch, try:

~~~~
git diff master..update-ipsets-commands -- sbin/update-ipsets.in
~~~~

**Comment by philwhineray** [comment]
I pushed a new script `packaging/detect-cmd.pl` to try to find any remaining plain command uses.

I will integrate it into the pre-commit hook soon but for now it can be run manually e.g.:

~~~~
./packaging/detect-cmd.pl configure.ac sbin/firehol.in
~~~~

I aslo fixed up the handful of problems I found with it.

**Comment by ktsaou** [comment]
Perfect.

A few more issues related to external commands:

1. `adnshost` is no longer needed (I have removed it from `update-ipsets` but it is still in autoconf)
2. `pv`, same as `adnshost`
3. `aggregate` in `link-balancer` should be replaced with `iprange` and removed from autoconf

In gentoo, I cannot run `./configure` as a normal user. It fails with this:

```sh
checking for ss... no
configure: error: cannot find required executable, bailing out
```

I have `ss` installed, but it is located as `/sbin/ss` which apparently is not searched for normal users.
With `sudo ./configure ...` it works.
It also works if I execute the following before `./configure`:

```sh
export PATH="${PATH}:/sbin"
```

Finally, I think `iprange` should be installed in `/usr/bin`, not `/usr/sbin`.

**Comment by philwhineray** [comment]
Good timing - my next intention was to put in place a check to ensure that the required programs list match what is actually used in each script - it will fit nicely with your changes.

I will fix the rest at the same time; for the configure problem it should be a simple fix: some items in configure.ac already have an optional extra path appended but I may just try to make /sbin and /usr/sbin part of the default since they will so often be needed. Putting iprange in /usr/bin may mean changing the layout of the repo to match; we will see.

**Comment by philwhineray** [comment]
I did everything except the aggregate to iprange change so far, including updating the checks.

**Comment by ktsaou** [comment]
Nice.

It is nothing to replace aggregate with iprange. Just remove the one and put the other. No command line options to either.

There is also a `tar` in `update-ipsets`.

**Comment by philwhineray** [comment]
OK, it's done.

**Comment by ktsaou** [comment]
@philwhineray I see that with `--prefix=/usr` the `FIREHOL_CONFIG_DIR` is set to `/usr/etc/firehol`.

Is this right?

**Comment by philwhineray** [comment]
I think so - autoconf doesn't seem to try to handle this case automatically (look at `./configure --help`). 
You need to add `--sysconfdir=/etc` explicitly.

**Comment by ktsaou** [comment]
ok, thanks!
